File: src/lib/payment/alipay.ts
import { AlipaySdk } from 'alipay-sdk';
import { paymentConfig, isAlipayConfigured } from './config';
import type { PaymentService, CreatePaymentParams, CreatePaymentResult, PaymentCallbackData } from './types';
import { createPaymentRecord, updatePaymentStatus, getPaymentByOutTradeNo } from './utils';

class AlipayPaymentService implements PaymentService {
  private alipay: AlipaySdk | null = null;

  constructor() {
    if (isAlipayConfigured()) {
      this.alipay = new AlipaySdk({
        appId: paymentConfig.alipay.appId,
        privateKey: paymentConfig.alipay.privateKey,
        alipayPublicKey: paymentConfig.alipay.alipayPublicKey,
        gateway: paymentConfig.alipay.sandbox 
          ? 'https://openapi.alipaydev.com/gateway.do' 
          : 'https://openapi.alipay.com/gateway.do',
      });
    }
  }

  async createPayment(params: CreatePaymentParams): Promise<CreatePaymentResult> {
    if (!this.alipay) {
      return { success: false, error: '支付宝支付未配置' };
    }

    try {
      const outTradeNo = `ALI${Date.now()}${Math.random().toString(36).substring(2, 8)}`;
      
      const result = this.alipay.pageExec(
        'alipay.trade.page.pay',
        {
          notify_url: paymentConfig.alipay.notifyUrl,
          return_url: paymentConfig.alipay.returnUrl,
          bizContent: {
            out_trade_no: outTradeNo,
            product_code: 'FAST_INSTANT_TRADE_PAY',
            total_amount: params.amount.toFixed(2),
            subject: `订阅计划-${params.planId}`,
          },
        }
      );

      const expiredAt = new Date();
      expiredAt.setHours(expiredAt.getHours() + 2);

      const payment = await createPaymentRecord({
        userId: params.userId,
        planId: params.planId,
        outTradeNo,
        method: 'alipay',
        amount: params.amount,
        payUrl: result,
        notifyUrl: params.notifyUrl,
        expiredAt,
        metadata: params.metadata,
      });

      return {
        success: true,
        paymentId: payment.id,
        outTradeNo,
        payUrl: result,
      };
    } catch (error) {
      console.error('创建支付宝支付失败:', error);
      return { success: false, error: '创建支付失败' };
    }
  }

  async handleCallback(data: any): Promise<PaymentCallbackData> {
    if (!this.alipay) {
      throw new Error('支付宝支付未配置');
    }

    try {
      const signVerified = this.alipay.checkNotifySign(data);
      
      if (!signVerified) {
        throw new Error('签名验证失败');
      }

      const outTradeNo = data.out_trade_no;
      const transactionId = data.trade_no;
      const amount = parseFloat(data.total_amount);
      const tradeStatus = data.trade_status;
      const paidAt = new Date(data.gmt_payment);

      if (tradeStatus === 'TRADE_SUCCESS' || tradeStatus === 'TRADE_FINISHED') {
        await updatePaymentStatus(outTradeNo, 'paid', {
          transactionId,
          paidAt,
        });

        return {
          outTradeNo,
          transactionId,
          amount,
          status: 'paid',
          paidAt,
        };
      } else {
        await updatePaymentStatus(outTradeNo, 'failed');
        
        return {
          outTradeNo,
          transactionId,
          amount,
          status: 'failed',
        };
      }
    } catch (error) {
      console.error('处理支付宝支付回调失败:', error);
      throw error;
    }
  }

  async queryPayment(outTradeNo: string): Promise<PaymentCallbackData | null> {
    if (!this.alipay) {
      throw new Error('支付宝支付未配置');
    }

    try {
      const result = await this.alipay.exec(
        'alipay.trade.query',
        {
          bizContent: {
            out_trade_no: outTradeNo,
          },
        }
      );

      const payment = await getPaymentByOutTradeNo(outTradeNo);
      if (!payment) {
        return null;
      }

      const tradeStatus = result.tradeStatus;
      let status: PaymentCallbackData['status'] = 'pending';

      if (tradeStatus === 'TRADE_SUCCESS' || tradeStatus === 'TRADE_FINISHED') {
        status = 'paid';
      } else if (tradeStatus === 'TRADE_CLOSED') {
        status = 'failed';
      }

      return {
        outTradeNo,
        transactionId: result.tradeNo,
        amount: parseFloat(result.totalAmount),
        status,
        paidAt: result.sendPayDate ? new Date(result.sendPayDate) : undefined,
      };
    } catch (error) {
      console.error('查询支付宝支付失败:', error);
      return null;
    }
  }

  async closePayment(outTradeNo: string): Promise<boolean> {
    if (!this.alipay) {
      throw new Error('支付宝支付未配置');
    }

    try {
      await this.alipay.exec(
        'alipay.trade.close',
        {
          bizContent: {
            out_trade_no: outTradeNo,
          },
        }
      );

      await updatePaymentStatus(outTradeNo, 'cancelled', {
        cancelledAt: new Date(),
      });

      return true;
    } catch (error) {
      console.error('关闭支付宝支付失败:', error);
      return false;
    }
  }
}

export const alipayPaymentService = new AlipayPaymentService();

File: backend/internal/payment/alipay.go
package payment

import (
	"context"
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"sort"
	"strconv"
	"strings"
	"time"
)

type AlipayProvider struct {
	config alipayConfiguration
	client *http.Client
}

func NewAlipayProvider(paymentConfig alipayConfiguration, client *http.Client) *AlipayProvider {
	if client == nil {
		client = &http.Client{Timeout: 10 * time.Second}
	}
	return &AlipayProvider{config: paymentConfig, client: client}
}

func (p *AlipayProvider) Method() PaymentMethod { return PaymentMethodAlipay }

func VerifyAlipaySignature(values url.Values, publicKeyPEM string) error {
	publicKey, err := parseRSAPublicKey(publicKeyPEM)
	if err != nil {
		return fmt.Errorf("alipay callback key: %w", err)
	}
	return verifyRSASignature(publicKey, canonicalAlipayValues(values), values.Get("sign"))
}

func canonicalAlipayValues(values url.Values) string {
	keys := make([]string, 0, len(values))
	for key := range values {
		if key != "sign" && key != "sign_type" {
			keys = append(keys, key)
		}
	}
	sort.Strings(keys)
	parts := make([]string, 0, len(keys))
	for _, key := range keys {
		parts = append(parts, key+"="+values.Get(key))
	}
	return strings.Join(parts, "&")
}

func (p *AlipayProvider) VerifyCallback(_ context.Context, request ProviderCallbackRequest) (ProviderCallback, error) {
	if err := VerifyAlipaySignature(request.Form, p.config.AlipayPublicKey); err != nil {
		return ProviderCallback{}, err
	}
	amount, err := strconv.ParseFloat(request.Form.Get("total_amount"), 64)
	if err != nil {
		return ProviderCallback{}, fmt.Errorf("parse alipay amount: %w", err)
	}
	status := PaymentStateFailed
	if request.Form.Get("trade_status") == "TRADE_SUCCESS" || request.Form.Get("trade_status") == "TRADE_FINISHED" {
		status = PaymentStatePaid
	}
	paidAt, _ := time.Parse("2006-01-02 15:04:05", request.Form.Get("gmt_payment"))
	transactionID := request.Form.Get("trade_no")
	return ProviderCallback{
		OutTradeNo: request.Form.Get("out_trade_no"), TransactionID: transactionID,
		Amount: amount, HasAmount: true, Currency: "CNY", Status: status, PaidAt: paidAt,
		CallbackKey: "alipay:" + transactionID,
	}, nil
}

func (p *AlipayProvider) Create(ctx context.Context, request ProviderCreateRequest) (ProviderCreateResult, error) {
	bizContent, err := json.Marshal(struct {
		OutTradeNo  string `json:"out_trade_no"`
		ProductCode string `json:"product_code"`
		TotalAmount string `json:"total_amount"`
		Subject     string `json:"subject"`
	}{OutTradeNo: request.OutTradeNo, ProductCode: "FAST_INSTANT_TRADE_PAY", TotalAmount: fmt.Sprintf("%.2f", request.Amount), Subject: "支付订单"})
	if err != nil {
		return ProviderCreateResult{}, fmt.Errorf("encode alipay payment: %w", err)
	}
	values := url.Values{
		"app_id":      {p.config.AppID},
		"method":      {"alipay.trade.page.pay"},
		"format":      {"JSON"},
		"charset":     {"utf-8"},
		"sign_type":   {"RSA2"},
		"timestamp":   {time.Now().Format("2006-01-02 15:04:05")},
		"version":     {"1.0"},
		"notify_url":  {request.NotifyURL},
		"return_url":  {p.config.ReturnURL},
		"biz_content": {string(bizContent)},
	}
	signature, err := p.sign(values)
	if err != nil {
		return ProviderCreateResult{}, err
	}
	values.Set("sign", signature)
	gateway := "https://openapi.alipay.com/gateway.do"
	if p.config.Sandbox {
		gateway = "https://openapi.alipaydev.com/gateway.do"
	}
	return ProviderCreateResult{PayURL: gateway + "?" + values.Encode()}, nil
}

func (p *AlipayProvider) Query(ctx context.Context, outTradeNo string) (ProviderStatus, error) {
	content, err := json.Marshal(struct {
		OutTradeNo string `json:"out_trade_no"`
	}{OutTradeNo: outTradeNo})
	if err != nil {
		return ProviderStatus{}, err
	}
	result, err := p.execute(ctx, "alipay.trade.query", content)
	if err != nil {
		return ProviderStatus{}, err
	}
	var payload struct {
		TradeQuery struct {
			TradeStatus string `json:"trade_status"`
			TradeNo     string `json:"trade_no"`
			TotalAmount string `json:"total_amount"`
			SendPayDate string `json:"send_pay_date"`
		} `json:"alipay_trade_query_response"`
	}
	if err := json.Unmarshal(result, &payload); err != nil {
		return ProviderStatus{}, fmt.Errorf("decode alipay query response: %w", err)
	}
	status := PaymentStatePending
	if payload.TradeQuery.TradeStatus == "TRADE_SUCCESS" || payload.TradeQuery.TradeStatus == "TRADE_FINISHED" {
		status = PaymentStatePaid
	} else if payload.TradeQuery.TradeStatus == "TRADE_CLOSED" {
		status = PaymentStateCancelled
	}
	amount, err := strconv.ParseFloat(payload.TradeQuery.TotalAmount, 64)
	if err != nil {
		return ProviderStatus{}, fmt.Errorf("parse alipay query amount: %w", err)
	}
	paidAt, _ := time.Parse("2006-01-02 15:04:05", payload.TradeQuery.SendPayDate)
	return ProviderStatus{TransactionID: payload.TradeQuery.TradeNo, Amount: amount, Currency: "CNY", Status: status, PaidAt: paidAt}, nil
}

func (p *AlipayProvider) Cancel(ctx context.Context, outTradeNo string) error {
	content, err := json.Marshal(struct {
		OutTradeNo string `json:"out_trade_no"`
	}{OutTradeNo: outTradeNo})
	if err != nil {
		return err
	}
	_, err = p.execute(ctx, "alipay.trade.close", content)
	return err
}

func (p *AlipayProvider) sign(values url.Values) (string, error) {
	privateKey, err := parseRSAPrivateKey(p.config.PrivateKey)
	if err != nil {
		return "", err
	}
	digest := sha256.Sum256([]byte(canonicalAlipayValues(values)))
	signature, err := rsa.SignPKCS1v15(rand.Reader, privateKey, crypto.SHA256, digest[:])
	if err != nil {
		return "", fmt.Errorf("sign alipay request: %w", err)
	}
	return base64.StdEncoding.EncodeToString(signature), nil
}

func (p *AlipayProvider) execute(ctx context.Context, method string, bizContent []byte) ([]byte, error) {
	values := url.Values{"app_id": {p.config.AppID}, "method": {method}, "format": {"JSON"}, "charset": {"utf-8"}, "sign_type": {"RSA2"}, "timestamp": {time.Now().Format("2006-01-02 15:04:05")}, "version": {"1.0"}, "biz_content": {string(bizContent)}}
	signature, err := p.sign(values)
	if err != nil {
		return nil, err
	}
	values.Set("sign", signature)
	gateway := "https://openapi.alipay.com/gateway.do"
	if p.config.Sandbox {
		gateway = "https://openapi.alipaydev.com/gateway.do"
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, gateway, strings.NewReader(values.Encode()))
	if err != nil {
		return nil, fmt.Errorf("create alipay request: %w", err)
	}
	request.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	response, err := p.client.Do(request)
	if err != nil {
		return nil, fmt.Errorf("alipay request: %w", err)
	}
	defer response.Body.Close()
	body, err := io.ReadAll(response.Body)
	if err != nil {
		return nil, fmt.Errorf("read alipay response: %w", err)
	}
	if response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusMultipleChoices {
		return nil, fmt.Errorf("alipay request returned status %d", response.StatusCode)
	}
	return body, nil
}
